File: medium/958_CheckCompletenessofaBinaryTree.py
# 958. Check Completeness of a Binary Tree
# medium
# https://leetcode.com/problems/check-completeness-of-a-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
        q = [(root, 0, 0)]
        curLvl = count = 0
        prevNum = -1
        while len(q) > 0:
            node, lvl, num = q.pop(0)
            if lvl == curLvl:
                count += 1
                if num != prevNum + 1:
                    return False
            else:
                if count != math.pow(2, curLvl):
                    return False
                if num != 0:
                    return False
                curLvl += 1
                count = 1

            prevNum = num         
            if node.left:
                q.append((node.left, curLvl + 1, num * 2))
            if node.right:
                q.append((node.right, curLvl + 1, num * 2 + 1))

        return True

File: medium/958_CheckCompletenessofaBinaryTree.go
// 958. Check Completeness of a Binary Tree
// medium
// https://leetcode.com/problems/check-completeness-of-a-binary-tree

package main
import (
	"container/list"
	"math"
)

type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
}

type nodeTrio struct {
    Node *TreeNode
    Lvl  int
	Left bool
}

func isCompleteTree(root *TreeNode) bool {
	q := list.New()
	q.PushBack(nodeTrio{Node: root, Lvl: 0, Left: true})

	curLvl, count := 0, 0
	prevLeft := false
	for q.Len() > 0 {
		front := q.Front()
		trio := front.Value
		q.Remove(front)

		if trio.Lvl == curLvl {
			count++
			if trio.Left == prevLeft {
				return false
			}
		} else {
			if float64(count) < math.Pow(2, float64(curLvl)) {
				return false
			}
			curLvl++
			count = 1
			prevLeft = true
		}

		if trio.Node.Left != nil {
			q.PushBack(nodeTrio{Node: trio.Node.Left, Lvl: curLvl + 1, Left: true})
		}
		if trio.Node.Left != nil {
			q.PushBack(nodeTrio{Node: trio.Node.Left, Lvl: curLvl + 1, Left: false})
		}
	}

	return true
}